docs/linters.md

```diff
@@ -0,0 +1,210 @@
+# Linters
+
+## Overview
+
+Zulip does extensive linting of much of its source code, including
+Python/JavaScript files, HTML templates (Django/handlebars), CSS files,
+JSON fixtures, Markdown documents, puppet scripts, and shell scripts.
+
+For some files we simply check for small things like trailing whitespace,
+but for other files, we are quite thorough about checking semantic
+correctness.
+
+Obviously, a large reason for linting code is to enforce the [Zulip
+coding standards](code-style.html).  But we also use the linters to
+prevent common coding errors.
+
+We borrow some open source tools for much of our linting, and the links
+below will direct you to the official documentation for these projects.
+
+- [jslint](https://github.com/douglascrockford/JSLint)
+- [mypy](http://mypy-lang.org/)
+- [puppet](https://puppet.com/) (puppet provides its own mechanism for validating scripts)
+- [pyflakes](https://pypi.python.org/pypi/pyflakes)
+
+Zulip also uses some home-grown code to perform tasks like validating
+indentation in template files, enforcing coding standards that are unique
+to Zulip, allowing certain errors from third party linters to pass through,
+and exempting legacy files from lint checks.
+
+## Running the linters
+
+If you run `./tools/test-all`, it will automatically run the linters (with
+one small exception: it does not run mypy against scripts).
+
+You can also run them individually:
+
+    ./tools/lint-all
+    ./tools/run-mypy
+    ./tools/run-mypy --scripts-only
+
+Finally, you can rely on our Travis CI setup to run linters for you, but
+it is good practice to run lint checks locally.
+
+Our linting tools generally support the ability to lint files
+individually--with some caveats--and those options will be described
+later in this document.
+
+We may eventually bundle `run-mypy` into `lint-all`, but mypy is pretty
+resource intensive compared to the rest of the linters, because it does
+static code analysis.  So we keep mypy separate to allow folks to quickly run
+the other lint checks.
+
+## General considerations
+
+Once you have read the [Zulip coding guidelines](code-style.html), you can
+be pretty confident that 99% of the code that you write will pass through
+the linters fine, as long as you are thorough about keeping your code clean.
+And, of course, for minor oversights, `lint-all` is your friend, not your foe.
+
+Occasionally, our linters will complain about things that are more of
+an artifact of the linter limitations than any actual problem with your
+code.  There is usually a mechanism where you can bypass the linter in
+extreme cases, but often it can be a simple matter of writing your code
+in a slightly different style to appease the linter.  If you have
+problems getting something to lint, you can submit an unfinished PR
+and ask the reviewer to help you work through the lint problem, or you
+can tap into other Zulip community resources like our Google mailing list.
```

Comment:
yeah maybe we should put it more obviously on http://zulip.readthedocs.io/en/latest/readme-symlink.html and then go from there